### ISSUE TITLE
GPU Memory exhausted (training)

### ISSUE THREAD

**Issue by vinamramunot-tech** [BenAAndrew/Voice-Cloning-App#7]
Label for this issue is help wanted.

# Hardware

Processor: Intel Core i9 - 9900K CPU @ 3.60GHz
Installed RAM: 64.0 GB
GPU: NVIDIA GeForce RTX 2070 - 8GB

---

# Attempt

![command](https://user-images.githubusercontent.com/56406734/111549929-7a5eca80-8753-11eb-8760-668301665f91.PNG)

---

# Failure

![gpu blasted](https://user-images.githubusercontent.com/56406734/111550647-e55cd100-8754-11eb-90ff-a6530363819d.PNG)

---
# Tried Steps
1. Epoch selection:
    - 8000
    - 5000
    - 3000
    - 1000 and 10 iterations

--- 
# Steps that might help to resolve

I found some articles that might be be helpful to resolve this:
1. https://pytorch.org/docs/stable/notes/faq.html
2. https://forums.fast.ai/t/clearing-gpu-memory-pytorch/14637/4

---
# Questions
1. Would you suggest to reduce the iterator counter from 1000 to a smaller number? If yes then what number would be ideal.
2. Would you suggest to reduce the epochs from 8000 to a smaller number? If yes then what number would be ideal.

**Comment by BenAAndrew** [comment]
Hi @vinamramunot-tech, The issue is to do with the batch size rather than epochs. To try and make the user experience as straightforward as possible I automatically calculate this value based on GPU memory in https://github.com/BenAAndrew/Voice-Cloning-App/blob/main/training/utils.py 

The formula I'm using is 
![image](https://user-images.githubusercontent.com/35925918/111623956-f9e7aa80-87e2-11eb-8c60-41e214e9e156.png)

In your case, this would result in a batch size of 28 ((8 - 1) * 4 = 28). Unfortunately, this a bit of an approximation and will be improved in a future development to take into account the dataset size. 

I think your issue is one of the following:
1. Your audio clips are too big (could you check the produced clips are 22050hz Mono 32kbs)
2. My batch size calculation hasn't worked (perhaps something else is using memory in your GPU)

Could you send me a screenshot of the output of `nvidia-smi` and also check the properties of your audio (I use an app called MediaInfo)

**Comment by vinamramunot-tech** [comment]
This is the info after running `nvidia-smi`

```    
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 456.71       Driver Version: 456.71       CUDA Version: 11.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name            TCC/WDDM | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce RTX 2070   WDDM  | 00000000:01:00.0  On |                  N/A |
|  0%   40C    P8    21W / 185W |   1516MiB /  8192MiB |      1%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A       348    C+G   ...icrosoft VS Code\Code.exe    N/A      |
|    0   N/A  N/A       932    C+G   ...udibleRT.WindowsPhone.exe    N/A      |
|    0   N/A  N/A      1268    C+G   Insufficient Permissions        N/A      |
|    0   N/A  N/A      1632    C+G   ...lPanel\SystemSettings.exe    N/A      |
|    0   N/A  N/A      2480    C+G   Insufficient Permissions        N/A      |
|    0   N/A  N/A      6380    C+G   ...cw5n1h2txyewy\LockApp.exe    N/A      |
|    0   N/A  N/A      7200    C+G   ...bbwe\Microsoft.Photos.exe    N/A      |
|    0   N/A  N/A      7780    C+G   C:\Windows\explorer.exe         N/A      |
|    0   N/A  N/A      8744    C+G   ...artMenuExperienceHost.exe    N/A      |
|    0   N/A  N/A      9112    C+G   ...w5n1h2txyewy\SearchUI.exe    N/A      |
|    0   N/A  N/A      9192    C+G   ...w5n1h2txyewy\SearchUI.exe    N/A      |
|    0   N/A  N/A      9624    C+G   ...ekyb3d8bbwe\YourPhone.exe    N/A      |
|    0   N/A  N/A     11148    C+G   ...me\Application\chrome.exe    N/A      |
|    0   N/A  N/A     11400    C+G   ...y\ShellExperienceHost.exe    N/A      |
|    0   N/A  N/A     12524    C+G   ...b3d8bbwe\WinStore.App.exe    N/A      |
+-----------------------------------------------------------------------------+

```

**Comment by vinamramunot-tech** [comment]
This is the output from the MediaInfo app for the `mp3` that I shared with you:

![audio-quall](https://user-images.githubusercontent.com/56406734/111665529-35857300-87e9-11eb-913f-abc8d53168e7.PNG)

**Comment by BenAAndrew** [comment]
@vinamramunot-tech Sorry, I meant the clips produced by dataset generation. You can find them in data\datasets\ _datasetname_ \wavs

**Comment by vinamramunot-tech** [comment]
This is the info on the produced wavs

![Capture](https://user-images.githubusercontent.com/56406734/111667242-e50f1500-87ea-11eb-9e31-5c16ad80bde5.PNG)

**Comment by vinamramunot-tech** [comment]
# Questions

1. If I may ask, what is the ideal setting of the audio that should be produced from mp3 to wav? 
2. Should reduce my dataset?
3. Should I try to change the batch size from 4 to 2?

I am apologize for rushing here, just trying to brain storm some solutions.

**Comment by BenAAndrew** [comment]
@vinamramunot-tech Hi, sorry for the late response. 

1. You have the same format I'd expect for the wavs produced (single channel, 16 bit, 352kb/s, 22050hz)
2. I wouldn't. More data is always ideal
3. It's odd that you'd need a batch size that small. As you have an 8GB card I'd expect a batch size of 20+. Could you try downloading the dataset I produced for David Attenborough from the voice sharing hub (https://voice-sharing-hub.herokuapp.com/) and let me know what the max batch size you can use to train on that is

**Comment by vinamramunot-tech** [comment]
@BenAAndrew 

1. Isn't 8000 a huge number for the epoch? Do you think reducing this would affect the quality of output?
2. Do you think reducing the batch size will affect the quality of the output?

I will try David Attenborough dataset to see what is the max batch I can use.

**Comment by BenAAndrew** [comment]
@vinamramunot-tech 

1. Yeah 8000 is pretty overkill. You can always start small (say 3000), test the results and then continue training if needed
2. No but it will take longer. I'd say 2 is too small (especially given you have 8GB GPU memory). Sounds like a problem with the data (maybe being too big) but we will know when you test with David Attenborough.
3. Yeah batch size 2 would take an extremely long time. Even when training with a larger batch size it will likely take at least a week to get good results. Using transfer learning may help speed this up.

**Comment by vinamramunot-tech** [comment]
@BenAAndrew Could this be possible that because I am using my dedicated graphics card for display it is not giving me full potential? Should I shift my display driver to integrated graphics?

Will update soon on the test for David Attenborough!

**Comment by BenAAndrew** [comment]
@vinamramunot-tech I wouldn't have thought you'd need to switch your display. You can check your GPU memory usage by running `nvidia-smi` and you should find 90%+ is available.  It is more likely an issue with the dataset your using

**Comment by vinamramunot-tech** [comment]
@BenAAndrew okay then will look into the dataset.

**Comment by BenAAndrew** [comment]
@vinamramunot-tech did you have a chance to try the David Attenborough dataset?

**Comment by BenAAndrew** [comment]
@vinamramunot-tech Closing this as these issues have been fixed since release 0.4

**Comment by vinamramunot-tech** [comment]
@BenAAndrew Apologies for the delay but was able to fix this. A lot of background processes were utilizing the memory which I didn't realize. Also the changes were able to fix this.
